### ISSUE TITLE
[Epic] adapter: improve migration framework

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#17466]
### Spec
The adapter migration framework needs some more thoughtful attention. We want to improve our confidence in the framework that it
a)  will work as intended
b) is able to support all the types of migrations we'd like to make, and
c) will prevent any catalog-breaking changes from being released.

Specifically, we should
1) Start by identifying the set of problems we wish to prevent and objectives of our migration framework, covering both stash and catalog migrations. Right now it's not clear and there's no explicit mandate. This is the most critical and difficult piece of work.
2) Once we have a clear understanding of objectives of our migration framework, compare that to the existing implementation(s) (which span 2+ places in the code).
3) Finally, either
    a) our assessment in step 2 above concludes that the existing implementation works, and we make minor improvements and add documentation
    b) we decide to redesign the migration framework, and kick off a formal design process.

Some ideas/related tickets:
- See contents below - change stash migrations to not use indexes into a vector to figure out where to start
- Integrate stash-debug into CI and/or release process workflows  (some similar ideas are being proposed for compute here https://github.com/MaterializeInc/materialize/issues/17883)
- https://github.com/MaterializeInc/materialize/issues/17824
- https://github.com/MaterializeInc/materialize/issues/14125
- https://github.com/MaterializeInc/materialize/issues/14126
  - This issue has a good overview of our current migration framework, but is not a high priority item itself.
- https://github.com/MaterializeInc/materialize/issues/17236
- https://github.com/MaterializeInc/materialize/issues/18485 (@mjibson  was working on this last I checked)
- https://github.com/MaterializeInc/materialize/issues/11435 (old tracker for initial implementation)

----
### Original ticket content 
We perform migrations in multiple places on envd start (in order):

- Stash migrations [link](https://github.com/MaterializeInc/materialize/blob/e96e7e2030de8eecd4a222d9c48514ddea5afc15/src/adapter/src/catalog/storage.rs#L567), called by envd's `serve` function
- Catalog migrations [link](https://github.com/MaterializeInc/materialize/blob/e96e7e2030de8eecd4a222d9c48514ddea5afc15/src/adapter/src/catalog.rs#L2584), called by Catalog::open, which is called by coord's `serve` function

Stash migrations are indexed, and we know which one we've performed, so we start at the one after and do those. Catalog migrations always run all catalog items. A problem with the existing migrations is that often (always?) need to be around for 1 release, and then can be deleted at some release after that. (A migration might exist in the future where a migration *must* be deleted before the next release.) We can usually cleanly delete the catalog migrations, but the stash migrations must stay around in their indexed place. Keeping around old migrations that won't ever do anything again can be bad because they force some unused data structures or fields to exist that should be removed.

Proposal from @benesch:

> I think just have each migration declare its migration ID rather than inferring it from its position in the array, and leave a comment saying what the next migration ID should be.
> In general we want to make these migrations idempotent if applied twice because if something goes wonky with the release (e.g. right now, v0.42 is at risk, and there’s a possibility we go straight from v0.41 -> v0.43), we might want to do something nonstandard with the version progression.
> So I think each migration carrying its migration ID, plus a comment saying the release at which a migration should/can be removed, is a good start? And then we can do a sweep and remove old migrations periodically, but probably don’t need to be too strict about removing them at exactly the next release?
> If we do write a migration that 100% needs to be removed after exactly one release: yeah enforcing that in code sounds good. But ideally that’s not the common case.

**Comment by mjibson** [comment]
Some goals I'd like (some discussed in huddle):

1. Compile-time errors if structs are changed in an unsafe way. Or maybe better said: it is not even possible to "just change the struct", instead everything must be written as a transformation from some previous state. Maybe some fancy build.rs thing that aggregates migrations and auto-generates the *current* struct, while it knows how to migrate to the current struct from any previous version.
2. Today when we change a struct we often will add a new migration about it AND have to modify the *initial* migration. Old users thus have a different upgrade path than new users. I'm worried that we'll sometimes diverge and have hard-to-find bugs. The recent stash bug for example required upgrading over 4 or so versions, but 2 was not enough. We should have a goal that we never (or maybe just are unable) to modify old migrations like that, so we never have these diverging data paths.

**Comment by parker-timmerman** [comment]
I'll write up a formal design doc, but was hoping to get some light adhoc feedback before that. As far as I can tell there are currently three kinds of migrations, Stash migrations, Catalog migrations, Builtin Object migrations. For now I'm going to focus on the first, Stash migrations.

### Stash Migrations
After looking through the code I mainly see two issues with our Stash migrations:

1. Migrations are provided a [`catalog::Transaction`](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/storage.rs#L1377), which in-turn contains [`TableTransaction`](https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/lib.rs#L777-L789)s which are [created using](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/storage.rs#L1309-L1375) [`TypedCollection`](https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/lib.rs#L352-L357)s. The `K` and `V` type args on `TableTransaction` and `TypedCollection` define how the underlying data is serde'd.
2. Migrations are a [list of closures](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/storage.rs#L84-L88) that we index into based on the user version we read from the stash.

[1] is a problem because a `catalog::Transaction` is what the rest of `adapter` uses to persist data into the catalog, which means it needs to have the latest/newest types. Using this for our migrations, means the types in a `catalog::Transaction` have to simultaneously support deserializing the current data from the stash, and support serializing whatever new data we want to add. This is what forces us into the pattern of doing two migrations to add a new field to the stash, one to add it as an `Option<T>`, then at least one release later, another to unwrap the field to just `T`. Improving this behavior is documented in https://github.com/MaterializeInc/materialize/issues/17824.

[2] is a problem because it makes our migrations fragile. We can't remove old migrations, instead we need to replace them with a no-op, otherwise our indexing would select the wrong closure.

### Proposed Changes
1. Maintain a `const CURRENT_STASH_VERSION: u64`. This helps us understand if the stash is "up-to-date" or whether we need to run migrations.
2. Structure migrations as a match statement, specifically:
```
fn migrate(...) {
  let user_version = // ... get the user version
  
  match user_version {
    0 => // todo migrate
    1 => // todo migrate
    ...
    // Nothing to do, we're up to date!
    CURRENT_STASH_VERSION => (),
    (CURRENT_STASH_VERSION + 1).. => panic!("Unrecognized stash version"),
  }
}
```
What this allows us to do is at build time, make sure we're handling migrations for all Stash versions. If you bump `CURRENT_STASH_VERSION` but don't specify a migration path, we'll fail to compile.

> Note: We can also use match guards, e.g. `0 if build.version() <= 0.46 =>` to make sure a migration only runs for certain releases.

3. Snapshot our Stash types. Maintaining old/current types of what used to/currently exists in the Stash allows us to safely open the Stash in a type-safe manner with `Type Old`, write easy to understand code converting `Type Old` to `Type New`, and then write back to the stash a serialized blob of `Type New`.

4. Move the migration down a layer, into the `stash` crate, and migrate using a `Stash` struct itself. The above described `fn migrate(...)` would take a `&mut Stash`. We can't use a `catalog::Transaction` to open the Stash with `Type Old`, and then later open it with `Type New`, we'd either have to make `catalog::Transaction` super generic, or duplicate a ton of code. I propose as part of [opening a Stash](https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/postgres.rs#L308) or immediately following open, we do the migration. The migration itself would operate with `TypedCollection`s. Specifically we'd add a new method to `TypedCollection` called `migrate(...)`, which looks something like:
```
pub enum MigrationAction<K, Kp, Vp> {
    Add(Kp, Vp),
    Delete(K),
    UpgradeKey(K, Kp),
    UpgradeValue(K, Vp),
    UpgradeEntry(K, (Kp, Vp)),
}

impl<K, V> TypedCollection<K, V> {
    pub fn migrate<Kp, Vp, F>(self, stash: &mut Stash, f: F) -> Result<(), ...>
    where
        F: FnOnce(&mut Stash, &TypedCollection<K, V>) -> Vec<Item = MigrationAction<K, Kp, Vp>>,
    {
        let changes = f(stash, &self);
        // TODO: iterate over all of the changes, creating an append batch
    }
}
```
Given the current `TypedCollection` and a Stash users then return a collection of changes, which we then run under a single transaction.

#### Snapshotting Types
IMO the most important thing to get correct here, is snapshotting our types, said another way, maintaining some record of what types we used to serialize in the Stash. If we do this incorrectly it can add a ton of overhead to our migrations, which makes them harder to understand and easier to get wrong.

An ideal snapshotting system would:
1. Be easy to understand and use
2. Run automatically, or with a single command
3. Detect when previous snapshots changed, ideally result in a build failure
4. Generate tests to make sure our types roundtrip through serialization (avoid future issues like https://github.com/MaterializeInc/materialize/pull/18418)

I believe we can achieve all of this with Rust procedural macros, specifically we define a macro that supports the following:

```
#[stash_object]
struct RoleValue {
    #[stash(removed = 10)]
    pub name: String,
    #[stash(added = 10)]
    pub role: SerializedRole,
}

#[stash_object]
#[stash(added = 10)]
pub struct SerializedRole {
    pub name: String,
    #[stash(added = 11)]
    pub attributes: RoleAttributes,
    #[stash(added = 12)]
    pub membership: RoleMembership,
}
```

We'd then generate the following structs:

```

// ... RoleValueV0 through RoleValueV8

struct RoleValueV9 {
  pub name: String
}

struct RoleValueV10 {
  pub role: SerializedRoleV10,
}

struct RoleValueV11 {
  pub role: SerializedRoleV11,
}

struct RoleValueV12 {
  pub role: SerializedRoleV12,
}

pub struct SerializedRoleV10 {
  pub name: String,
}

pub struct SerializedRoleV11 {
  pub name: String,
  pub attributes: RoleAttributesV11,
}

pub struct SerializedRoleV12 {
  pub name: String,
  pub attributes: RoleAttributesV12,
  pub membership: RoleMembershipV12,
}
```

For every version between 0 and `CURRENT_STASH_VERSION` we generate a struct for each object in the stash, aliasing the latest versions, e.g. `pub type RoleValue = RoleValueV12`. Also (for better or worse) proc macros are allowed to do anything, including reading and writing files from the filesystem. So while generating code, we could also read in saved snapshots of previous versions, to make sure we didn't accidentally change them.

tl;dr we use proc macros to describe the changes that occur for a type, which then generates a struct for each version of the DB, asserts previous versions didn't change based on persisted snapshots, and all of this happens as part of `cargo build`.

**Comment by jkosh44** [comment]
At a high level it looks really good. I have some coments:

> Maintain a const CURRENT_STASH_VERSION: u64. This helps us understand if the stash is "up-to-date" or whether we need to run migrations.

If possible maybe we want to re-use the version number here. That would at least reduce the overhead of adding another number that needs to be incremented every release.

> Structure migrations as a match statement, specifically:...

When adding a migration, generally the reason falls into two categories:
1. This migration is only for existing deployments, i.e. someone who would have a non-zero stash version number. An example of this was adding a default owner to all objects that existed before the concept of owners (https://github.com/MaterializeInc/materialize/blob/89fc884ef5178a429c7a1103a6ae8512b3e50130/src/adapter/src/catalog/storage.rs#L413-L419).
2. This migration is for existing deployments, i.e. someone who would have a non-zero stash version number, AND for new deployments, i.e. someone who would have a zero stash version number. An example of this was when we added the PUBLIC psuedo role (https://github.com/MaterializeInc/materialize/blob/89fc884ef5178a429c7a1103a6ae8512b3e50130/src/adapter/src/catalog/storage.rs#L370-L375).

It seems like this format would easily allow us to do type (1) migrations, we just need to make sure it also allows us to do type (2) migrations.

> Snapshot our Stash types. 

I really like this idea. I would be interested to see what it would look like to use one of these structs though, because I'm not very familiar with proc macros.

**Comment by mjibson** [comment]
> pub fn migrate<Kp, Vp, F>(self, stash: &mut Stash, f: F) -> Result<(), ...>

This should probably take `tx: &mz_stash::Transaction` instead of a Stash, so that multiple TypedCollections can be modified in a single transaction.

What about a case where a migration needs to consult the values of items from two separate collections? The proposed code looks pretty tied to a single collection.

One benefit of using a `storage::Transaction` is it uses a bunch of `mz_stash::TableTransaction` which come with uniqueness violation checks. Unclear if we need that feature during migrations, but notable to lose it.

I really like the proc macro! Some questions:
- should every field require an added attr? (I think yes, even for v0, just for clarity.)
- after adding a removed attr (and it'd still have to have added even in that case), does this mean we cannot ever re-use that name? (this is probably fine, just thinking through it.)
- This would mean that the actual migrations must never reference the type aliased (unversioned, most recent) struct types, right? Is there a way to enforce this automatically?
- I don't see anything calling out specifically how or where migrations between struct versions would occur. Would migrations need to know that they must call UpgradeX and pass some function to migration from a previous version to the current version?
- It seems like we'd have enough info to know exactly when a struct migration was required. Could the proc macro...be smart enough to require those functions exist?

It's not clear to me whether this means we can periodically squash old migrations or not. We can absolutely assume that the current release is, at the most, getting run from a stash from last week's release. There are at most two stash versions on disk in production at any time: this week's and last week's.

**Comment by parker-timmerman** [comment]
Thanks @jkosh44 and @mjibson for your feedback! I think I got to all your questions, let me know if you have any more :)

> If possible maybe we want to re-use the version number here. That would at least reduce the overhead of adding another number that needs to be incremented every release.

This "stash version number" already implicitly exists with the list of migrations that we currently have. Making it explicit wouldn't require us to bump it every release, i.e. v51 and v52 of Materialize can both have "stash v13". I believe de-coupling them also de-risks if we ever miss a release, e.g. upgrading from v51 straight to v53, since we wouldn't rely on the build number, missing a release shouldn't matter.

> It seems like this format would easily allow us to do type (1) migrations, we just need to make sure it also allows us to do type (2) migrations.

That's a good call out! I looked through our existing migrations and I believe they should all work with this model. At a minimum we can bump the `CURRENT_STASH_VERSION` to indicate something has changed, even if the types themselves have not. This should allow us to handle type (2) migrations.

> I would be interested to see what it would look like to use one of these structs though, because I'm not very familiar with proc macros.

Using these structs should just be like using the existing types today, proc macros essentially just do code generation, and in this case they'd generate snapshots of the previous versions that were stored in the stash.

> This should probably take tx: &mz_stash::Transaction instead of a Stash, so that multiple TypedCollections can be modified in a single transaction.

Sweet, thanks for the feedback, I wasn't sure what the right "connection" type would be. Using an `mz_stash::Transaction` I think makes the most sense!

> What about a case where a migration needs to consult the values of items from two separate collections? The proposed code looks pretty tied to a single collection.

This should work just fine, what I envisioned was something like:
```
const COLLECTION_A = ... ;
const COLLECTION_B = ... ;

let candidates = vec![];
for entry in COLLECTION_B.iter(...) {
  candidates.push(...);
}

COLLECTION_A.migrate(|stash, col_a| {
  let changes = vec![];
  for entry in COLLECTION_A.iter(...) {
    // Check based on candidates from COLLECTION_B.
    if candidates.contaions(entry) {
      changes.append(...);
    }
  };
  changes
})
```
You should be able to use I was thinking that we can manually enforce a style for migrations that in the first few lines you define the collections you're going to use.

> One benefit of using a storage::Transaction is it uses a bunch of mz_stash::TableTransaction which come with uniqueness violation checks. Unclear if we need that feature during migrations, but notable to lose it.

Yeah I wasn't sure if we'd need those uniqueness violation checks, if we do, I think it should be possible to use `TableTransaction` instead of `TypeCollection` for migrations, but I'd need to play around with it a bit more to be sure.

#### Proc Macros
> should every field require an added attr? (I think yes, even for v0, just for clarity.)

We could, or any fields without it could implicitly get "added in" the same version their parent struct was. If you add a new field to an existing struct and don't specify the `#[stash(added = ...)]` attribute, we'd fail our build check as we'd detect previous versions of the struct as having changed.

> after adding a removed attr (and it'd still have to have added even in that case), does this mean we cannot ever re-use that name? (this is probably fine, just thinking through it.)

We could re-use the same, in fact we'd need to support this for the case of converting something to a new type, e.g.
```
#[stash_object]
struct Role {
  #[stash(removed = 10)]
  id: u64,
  #[stash(added = 10)]
  id: String,
}
```
Is at least _a way_, to support re-typing a field. We could also support a more explicit "change" or "retype" attribute if we want, but that might get a bit complicated.

> This would mean that the actual migrations must never reference the type aliased (unversioned, most recent) struct types, right? Is there a way to enforce this automatically?

Correct, migrations must never reference the type aliased "latest" types, because those would change making the migrations incorrect. A way to prevent this would at least be to have two crates like `stash-latest` and `stash-objects` and enforce with `deny.toml` we import one but not the other, but that's a bit verbose. We could make a lint that checks our imports, but we miiiiiight be able to do something tricky with module visibility, not sure though.

> I don't see anything calling out specifically how or where migrations between struct versions would occur. Would migrations need to know that they must call UpgradeX and pass some function to migration from a previous version to the current version?
It seems like we'd have enough info to know exactly when a struct migration was required. Could the proc macro...be smart enough to require those functions exist?

I think we could define some trait like:
```
trait Upgrade<T> {
  fn upgrade(self) -> <todo>;
}
```
And then enforce that trait to be implemented for the types the proc macro determine need a migration. Then instead of passing a closure to `TypedCollection::migrate` like I define above, we can call this trait.

So we can force folks to implement the trait, but I don't know if we can automatically call it for them, while still being flexible enough to support all different kinds of migrations. For example, if to migrate one type to another you need to reference multiple collections, I'm not sure if we can make the `Upgrade` trait generic enough to support that. 

I definitely want to get to the behavior you're describing. My hope was to implement an initial version of these new migrations to improve the current flow, and then iterate towards this north star. I think the current proposal allows us to do that.

> It's not clear to me whether this means we can periodically squash old migrations or not.

I don't think we can squash old migrations, not at least until we have some sort of "initialization step". For example, the [first migration](https://github.com/MaterializeInc/materialize/blob/89fc884ef5178a429c7a1103a6ae8512b3e50130/src/adapter/src/catalog/storage.rs#L90) populates tables with a bunch of initial values, so we need to retain it. If we were to introduce a separate "initialization" code path, then we could probably squash old migrations.

**Comment by parker-timmerman** [comment]
### Builtin Migrations

Within the catalog there exists [Builtin objects](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/builtin.rs#L51), these are predefined tables, views, functions, etc. that each user has. An example would be our `mz_catalog` tables. For each of these builtin objects we define a [fingerprint](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/builtin.rs#L182-L199), and if this fingerprint ever changes then we determine the object as needing to be migrated. To migrate these objects, we drop the object and any dependents, and then recreate everything.

#### Problems
This migration path currently works for all objects, but has a few issues:

1. https://github.com/MaterializeInc/materialize/issues/14126. When renaming an object, all we really need to do is update the [mapping of name to object id](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/storage.rs#L2548), nothing needs to get dropped an recreated. But today because the [fingerprint for a `BuiltinView` is the raw sql](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/builtin.rs#L227), changing the name of an object requires us to update any `BuiltinView`s that depend on it, thus changing the fingerprint, and thus dropping all dependents, which is very inefficient.
2. https://github.com/MaterializeInc/materialize/issues/18767. If we ever drop and recreate a `Sink`, we'll re-emit previous data when it's recreated, which breaks our exactly-once semantics. If we migrate a Builtin object then we have the possibility of dropping and recreating a `Sink` that a user may have created.
3. https://github.com/MaterializeInc/materialize/issues/18763. Migrating a `BuiltinSource` is lossy, when we drop and recreate the source we lose all of the historical values. This isn't awful, but is something we'd like to improve.
4. https://github.com/MaterializeInc/materialize/issues/17236. Builtin migrations currently aren't tested very well. We should reintroduce tests we use to have for them, and possibly build out a more extensive test framework.

#### Proposals

1. Change `BuiltinView`s to store a SQL template, with objects that we depend on being variables in the template. This should prevent the fingerprint from changing when a dependent's name changes, and allows us to statically determine what objects a view depends on. For example it could look something like:
```
pub const MZ_DATAFLOW_OPERATORS: BuiltinView = BuiltinView {
    name: "mz_dataflow_operators",
    schema: MZ_INTERNAL_SCHEMA,
    sql: Template::new("CREATE VIEW mz_internal.mz_dataflow_operators AS
SELECT id, name
FROM {1}
WHERE worker_id = 0"
    ),
    dependencies: [
        (MZ_INTERNAL_SCHEMA, MZ_DATAFLOW_OPERATORS_PER_WORKER),
    ],
};
```
> Note: From a quick Google search String template-ing in Rust seems not highly used, so our mileage may vary here

2. To improve testing, in the immediate term we should complete https://github.com/MaterializeInc/materialize/issues/17236. Longer term snapshotting user Catalogs would go a long way for testing, e.g. [Poor man's Snowtrail](https://www.notion.so/materialize/Poor-man-s-Snowtrail-1c36cdc61a224948b623318a02db3bb8?pvs=4). We don't need all of the features that would be part of a "Snowtrail" though. To get meaningful data to test migrations what we'd need is, the ability to periodically and safely dump users' catalog (e.g. `cron` + `stash-debug`), a place to put these dumps (e.g. S3), and a test runner that can read from our storage, and start envd.
3. As described in https://github.com/MaterializeInc/materialize/issues/18767, we should prevent Sinks from referring to system objects. This mitigates the risk if a Builtin migration dropping a Sink. More investigation is needed to determine how/if we could migrate something a Sink depends on, without dropping the Sink itself.

**Comment by jkosh44** [comment]
> Change BuiltinViews to store a SQL template, with objects that we depend on being variables in the template. This should prevent the fingerprint from changing when a dependent's name changes, and allows us to statically determine what objects a view depends on.

Completing this issue: https://github.com/MaterializeInc/materialize/issues/17892#issuecomment-1505607071 would also prevent the fingerprint from changing when a dependent's name changes. I still like the templating idea though for identifying and modifying dependencies.

> To improve testing, in the immediate term we should complete https://github.com/MaterializeInc/materialize/issues/17236.

Just giving you a heads up, I think this is going to be a difficult task. It's unclear how we can trigger a built-in migration in a test, since the fingerprints are compiled into the binary. One idea is to have the test modify the stored fingerprints in the stash. This would trick Materialize into thinking there was a migration even though there wasn't really one. The issue with that approach is that it's not testing a realistic scenario, so there may be a better approach out there.
